### ISSUE TITLE
snabbvmx does not honor mtu setting

### ISSUE THREAD

**Issue by chrgraf** [Igalia/snabbswitch#487]
Dear all,

when testing vmxlwaftr I am facing issues with the MTU setting.

Both vmxlwaftr and snabbvmx ignore the 9000 MTU setting (junos config at end of this report) and encapsulate IPv4 frames with a size of 9042byte
I checked with lwaftr standalone and it behaves as expected - bytes larger as the configured 9000MTU are being dropped.
# topology

ubuntu2-90:e2:ba:94:2a:bc---------02:cf:69:15:81:01 vmxlwaftr
# the packet

13:39:20.896417 90:e2:ba:94:2a:bc > 02:cf:69:15:81:01, ethertype IPv4 (0x0800), length 9042: 10.0.1.100.17 > 10.10.0.0.1024: UDP, length 9000

[ingress.pcap.gz](https://github.com/Igalia/snabb/files/530287/ingress.pcap.gz)
# running snabbvmx standalone

```
root@ubuntu1:~/vmxlwaftr/tests/mtu-issue# usr/local/bin/snabb snabbvmx lwaftr --conf snabbvmx-lwaftr-xe1.cfg --id xe1 --pci 0000:81:00.1 --mac 02:cf:69:15:81:01
Ring buffer size set to 2048
loading compiled binding table from ./binding_table.txt.s.o
compiled binding table ./binding_table.txt.s.o is up to date.
Hairpinning: no
nic_xe1 ether 02:cf:69:15:81:01
IPv6 fragmentation and reassembly: no
IPv4 fragmentation and reassembly: no
lwAFTR service: enabled
Running without VM (no vHostUser sock_path set)
nh_fwd6: cache_refresh_interval set to 0 seconds
nh_fwd6: static next_hop_mac 90:e2:ba:94:2a:bc
loading compiled binding table from ./binding_table.txt.s.o
compiled binding table ./binding_table.txt.s.o is up to date.
nh_fwd4: cache_refresh_interval set to 0 seconds
nh_fwd4: static next_hop_mac 90:e2:ba:94:2a:bc
Ingress drop monitor: flush (threshold: 100000 packets; wait: 20 seconds; interval: 1.00 seconds)
```
# as seen, the IPv4 frame gets encapsulated by lwaftr and sent to tunnel - while it shall have been dropped due of its size larger as configured value

```
13:44:51.236419 90:e2:ba:94:2a:bc > 02:cf:69:15:81:01, ethertype IPv4 (0x0800), length 9042: 10.0.1.100.17 > 10.10.0.0.1024: UDP, length 9000
13:44:51.237671 02:cf:69:15:81:01 > 90:e2:ba:94:2a:bc, ethertype IPv6 (0x86dd), length 9082: 2a02:587:f700::100 > 2a02:587:f710::40: 10.0.1.100.17 > 10.10.0.0.1024: UDP, length 9000
```
# config-file snabbvmx - is here the mtu setting missing??

```
root@ubuntu1:~/vmxlwaftr/tests/mtu-issue# cat snabbvmx-lwaftr-xe1.cfg
return {
  lwaftr = "snabbvmx-lwaftr-xe1.conf",
  settings = {
  },
  ipv6_interface = {
    ipv6_address = "",
    cache_refresh_interval = 0,
    fragmentation = false,
    next_hop_mac = "90:e2:ba:94:2a:bc",
  },
  ipv4_interface = {
    ipv4_address = "192.168.5.2",
    cache_refresh_interval = 0,
    fragmentation = false,
    next_hop_mac = "90:e2:ba:94:2a:bc",
  },
}
```
# Now using same frame and test against lwaftr 

launching lwaftr

```
root@ubuntu1:~/vmxlwaftr/tests/mtu-issue# usr/local/bin/snabb  lwaftr run --conf snabbvmx-lwaftr-xe1.conf --on-a-stick 0000:81:00.1
Warning: No NUMA memory affinity.
Pass --cpu to bind to a CPU and its NUMA node.
loading compiled binding table from ./binding_table.txt.s.o
compiled binding table ./binding_table.txt.s.o is up to date.
```
# as seen in below dump lwaftr does not encapsulate the 9042@l2 large packet, which is the correct behavior

```
13:50:39.900392 90:e2:ba:94:2a:bc > 02:cf:69:15:81:01, ethertype IPv4 (0x0800), length 9042: 10.0.1.100.17 > 10.10.0.0.1024: UDP, length 9000
13:50:39.912371 90:e2:ba:94:2a:bc > 02:cf:69:15:81:01, ethertype IPv4 (0x0800), length 9042: 10.0.1.100.17 > 10.10.0.0.2048: UDP, length 9000
```
# config lwaftr

```
root@ubuntu1:~/vmxlwaftr/tests/mtu-issue# cat snabbvmx-lwaftr-xe1.conf
hairpinning = false,
vlan_tagging = false,
binding_table = binding_table.txt.s,
aftr_ipv6_ip = 2001:db8::1,
aftr_ipv4_ip = 192.168.5.2,
aftr_mac_inet_side = 02:cf:69:15:81:01,
inet_mac = 90:e2:ba:94:2a:bc,
aftr_mac_b4_side = 02:cf:69:15:81:01,
next_hop6_mac = 90:e2:ba:94:2a:bc,
ipv4_mtu = 9000,
ipv6_mtu = 9000,
policy_icmpv4_incoming = allow,
policy_icmpv4_outgoing = allow,
```
# snabb-binary

[snabb.gz](https://github.com/Igalia/snabb/files/530324/snabb.gz)
# junos cli softwire-config - with mtu 9000 setting

```
lab@lwaftr1# top show ietf-softwire:softwire-config
binding {
    br {
        br-instances {
            br-instance 0 {
                softwire-num-threshold 4000000;
                tunnel-payload-mtu 9000;
                tunnel-path-mru 9000;
                jnx-aug-softwire:ipv4_address 192.168.4.2;
                jnx-aug-softwire:fragmentation false;
                jnx-aug-softwire:cache_refresh_interval 1;
                jnx-aug-softwire:icmpv6_rate_limiter_n_packets 300000;
                jnx-aug-softwire:icmpv6_rate_limiter_n_seconds 4;
                jnx-aug-softwire:policy_icmpv4_incoming allow;
                jnx-aug-softwire:policy_icmpv4_outgoing allow;
            }
            br-instance 1 {
                softwire-num-threshold 4000000;
                tunnel-payload-mtu 9000;
                tunnel-path-mru 9000;
                jnx-aug-softwire:ipv4_address 192.168.5.2;
                jnx-aug-softwire:fragmentation false;
                jnx-aug-softwire:cache_refresh_interval 1;
                jnx-aug-softwire:policy_icmpv4_incoming allow;
                jnx-aug-softwire:policy_icmpv4_outgoing allow;
            }
        }
        jnx-aug-softwire:binding-table-file binding_table.txt;
    }
}
```

kindly have a look into it

thanks

christian

**Comment by mwiget** [comment]
I introduced this regression by setting the physical interface mtu to DEFAULT_MTU = 9500. This was done under the (wrong) assumption, that ipv4_mtu and ipv6_mtu will be honored by the lwaftr app. Though turns out, that is done by the fragmentation app only, which is mandatory for the lwaftr program, but optional for snabbvmx.

I'm working on a pull request to automatically calculate the MTU for the interface based on the ipv4_mtu and ipv6_mtu, by taking the numerically larger one and add 14 to cover the size of the ethernet header. The change will also print out the mtu to STDOUT together with the interface name.

**Comment by dpino** [comment]
Fixed in #488.

**Comment by chrgraf** [comment]
Many Thanks Diego

Christian

---

Christian Graf
Senior Consulting Engineer
Telephone +49 5651/336174
Mobile +49 175/2929 167
cgraf@juniper.net
www.juniper.net

Juniper Networks GmbH
Sitz der Gesellschaft / Registered office: Oskar-Schlemmer-Str. 15, 80807 München
Registergericht / Commercial registry: München, HRB 205571
Geschäftsführer / Directors: Johannes Albers, Stephen Byrne, Terrance Floyd Spidell

---

This email contains material that is confidential. The content of this email is for the sole use of the intended recipient(s).
Any review or distribution by persons other than the intended recipient(s) without the express permission of Juniper Networks is strictly prohibited.
If you are not the intended recipient, please contact the sender and delete/destroy all copies of this email and any related attachments.
Juniper does not guarantee the accuracy or completeness of third party materials or information
